UI, controller and suite.xml code for case search

kaapstorm opened the pull request:
This PR replaces the original review PR #11050 , now based off `master`. 

Don't merge yet.

This PR has grown to be ridiculously big. I'll break it into bite-size chunks when it's ready for merging. I still need to add some tests, and pull out commits that are kind of stand-alone, like ba219763

@proteusvacuum cc @millerdev

kaapstorm committed:
Add toggle

kaapstorm committed:
view stubs

kaapstorm committed:
suite.xml models

kaapstorm committed:
case search models

kaapstorm committed:
case search models

kaapstorm committed:
SyncRequestContributor

kaapstorm committed:
First simple test

kaapstorm committed:
Use id_strings for localization

kaapstorm committed:
Refer docstrings to sync-request defn

kaapstorm committed:
case claim view uses post_case_blocks

kaapstorm committed:
CaseSearch model must support multiple languages

kaapstorm committed:
claim case test

kaapstorm committed:
case search models: constants for suite.xml

kaapstorm committed:
update test

kaapstorm committed:
Update SyncRequestContributor based on feedback

kaapstorm committed:
search view uses CaseSearchConfig and CaseDBFixture

kaapstorm committed:
search view uses CaseSearchConfig and CaseDBFixture

kaapstorm committed:
get a deserialized instance from serialized data

kaapstorm committed:
Fixed CaseDBFixture instantiation

kaapstorm committed:
Case search config lets you set the command label

kaapstorm committed:
Use CaseTypeXpath with instance name

kaapstorm committed:
Tests support both DB backends

proteusvacuum committed:
return fixture as a string

proteusvacuum committed:
Flatten search results to be used in serializers

kaapstorm committed:
Case search endpoint test added

kaapstorm committed:
project settings case search config ui

kaapstorm committed:
view added

kaapstorm committed:
add get_or_none method to CaseSearchConfig model

kaapstorm committed:
view passed current values

kaapstorm committed:
initialize form values

kaapstorm committed:
post stub, drop bad imports

kaapstorm committed:
Manage properties better

kaapstorm committed:
view post

kaapstorm committed:
use Python naming convention where convenient

kaapstorm committed:
jslint

kaapstorm committed:
Only post "case_type" and "properties" fuzzy property attributes

kaapstorm committed:
Misplaced comment

kaapstorm committed:
caseTypes doesn't need to be observable because it doesn't change

kaapstorm committed:
create CaseSearchConfigJSON instance correctly

kaapstorm committed:
No need for a wrapped view model

kaapstorm committed:
Handle zero-length fuzzy properties

kaapstorm committed:
Pass is_search_enabled to template

kaapstorm committed:
searchViewModel

kaapstorm committed:
module UI

kaapstorm committed:
No point in returning more than 10 results

kaapstorm committed:
Appeasing dimagimon

kaapstorm committed:
Save POSTed search properties

kaapstorm committed:
Initialize view model with current values

kaapstorm committed:
FB223512: Iterate properties, not search_config

dimagimon commented on corehq/apps/ota/tests/test_search_claim_endpoints.py:
```diff
@@ -0,0 +1,60 @@
+from corehq.apps.domain.shortcuts import create_domain
+from corehq.apps.users.models import CommCareUser
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.form_processor.tests import run_with_all_backends
```
F401 'run_with_all_backends' imported but unused

kaapstorm committed:
Use CaseAccessors

dimagimon commented on corehq/apps/ota/views.py:
```diff
@@ -1,18 +1,30 @@
+from corehq.form_processor.serializers import CommCareCaseSQLSerializer, get_instance_from_data
+from corehq.pillows.const import MAX_SEARCH_RESULTS
 from corehq.tabs.tabclasses import ProjectSettingsTab
+from corehq.form_processor.models import CommCareCaseSQL
```
F401 'CommCareCaseSQL' imported but unused

kaapstorm committed:
Move claim_case into its own function

(cleanup.py might not be completely inappropriately named if you consider that it allows us to clean up views.py ;-) )

kaapstorm committed:
Make property names observables

kaapstorm committed:
Improved wording slightly

kaapstorm committed:
Fix unpacking request.POST keys

dimagimon commented on corehq/apps/domain/views.py:
```diff
@@ -2071,6 +2080,59 @@ def post(self, request, *args, **kwargs):
+                    fuzzy_dict[i][attr] = query_dict.getlist(key) if is_list else query_dict[key]
+            if not fuzzy_dict:
+                return []
+            return [fuzzy_dict[i] for i in range(max(fuzzy_dict.keys()) + 1) if fuzzy_dict[i]]
```
F812 list comprehension redefines 'i' from line 2107

kaapstorm committed:
Don't nest search_properties

kaapstorm commented:
@proteusvacuum not sure what I've added since you last glanced. Maybe :tropical_fish: from "No point in returning more than 10 results" (f17471b). Recent ones are just bug fixes.

kaapstorm committed:
Don't offer search if it isn't configured

kaapstorm committed:
Keep with instance('foo')/foo convention

kaapstorm committed:
Add search command & test

dimagimon commented on corehq/apps/app_manager/tests/test_suite_sync_request.py:
```diff
@@ -0,0 +1,42 @@
+from mock import patch
+
+from corehq.apps.app_manager.const import APP_V2
+from corehq.apps.app_manager.models import Application, Module, CaseSearch, CaseSearchProperty, UpdateCaseAction
```
F401 'UpdateCaseAction' imported but unused

proteusvacuum commented on corehq/apps/ota/views.py:
```diff
@@ -81,28 +77,8 @@ def claim(request, domain):
-        ).as_xml()
-        post_case_blocks([claim_case_block], {'domain': domain})
+        claim_case(domain, couch_user.user_id, request.POST['case_id'],
+                   host_type=request.POST.get('case_type'), host_name=request.POST.get('case_name'))
```
:+1:

proteusvacuum commented on corehq/apps/domain/views.py:
```diff
@@ -2097,23 +2097,25 @@ def unpack_fuzzies(query_dict):
+                    fuzzy_dict[i][attr] = query_dict.getlist(key) if is_list else query_dict[key]
             if not fuzzy_dict:
                 return []
             return [fuzzy_dict[i] for i in range(max(fuzzy_dict.keys()) + 1) if fuzzy_dict[i]]
```
I think you could achieve the same thing here more pythonically with `[item for key, item in fuzzy_dict.iter_items() if item]`, unless I've misunderstood what this is doing

kaapstorm replied:
Hmmm. Not sure why I did it that way -- maybe a little too focused on that index. Could even just `[item for item in fuzzy_dict.iter_values() if item]` or `filter(lambda x: x, fuzzy_dict.values())`. Thank you.

proteusvacuum commented:
Looks good Norman. Build fail, and conflicts, but looking forward to getting this in!
```
======================================================================

ERROR: test_case_list_form_media (corehq.apps.app_manager.tests.test_media_suite.LocalizedMediaSuiteTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/mnt/corehq/apps/app_manager/tests/test_media_suite.py", line 228, in test_case_list_form_media

    self._assert_app_strings_available(app, 'en')

  File "/mnt/corehq/apps/app_manager/tests/test_media_suite.py", line 266, in _assert_app_strings_available

    app_strings = commcare_translations.loads(app.create_app_strings(lang))

  File "/mnt/corehq/apps/app_manager/models.py", line 4744, in create_app_strings

    return gen.create_app_strings(self, lang)

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 161, in create_app_strings

    for part in self.app_strings_parts(app, lang, for_default=for_default):

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 250, in app_strings_parts

    for_default=for_default)

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 154, in create_custom_app_strings

    custom = dict(_create_custom_app_strings(app, lang, for_default=for_default))

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 115, in _create_custom_app_strings

    yield id_strings.search_property_locale(module, prop.name), trans(prop.label)

AttributeError: 'str' object has no attribute 'name'
```
(Same error in other build too)

kaapstorm committed:
Use module_offers_search

kaapstorm committed:
OCD

proteusvacuum commented:
:boot: :construction_worker:

kaapstorm commented:
:white_check_mark: 

Three things to note about this branch that I would like to add in the very near future, and all pretty small:

1. Save button on page to enable search and configure fuzzy properties, under Project Settings, isn't notified of changes. It is always active. Would be nice to enable after change, and disable after save.
2. When claiming a case, session doesn't store whether the user has already just claimed the case, as per spec. This means that multiple claims can be created for the same case by the same user. I should prevent this, and add a test for it.
3. Module case list page should allow user to change wording of the "Search" command. suite.xml allows for it, but UI doesn't offer it.

Dropping "Hold off on merge" but noting that these three things would be very nice, and (2) is required.

kaapstorm committed:
Prevent duplicate claims

kaapstorm committed:
Unpack GET params correctly

kaapstorm committed:
Catch NotFoundError

kaapstorm committed:
Break the tests

proteusvacuum commented:
:tada:
